scipy/interpolate/regrid_python.py

```diff
@@ -0,0 +1,1174 @@
+
+        if x.shape != y.shape:
+            x, y = np.broadcast_arrays(x, y)
+
+        if x.size == 0:
+            return np.zeros(x.shape + trailing, dtype=ndbs.c.dtype)
+        xi = np.stack((x.ravel(), y.ravel()), axis=-1)
+        vals = ndbs(xi, nu=(dx, dy), extrapolate=ndbs.extrapolate)
+        return vals.reshape(x.shape + trailing)
+
+def return_NdBSpline(fp, tck, degrees):
+    """
+    Build a 2D ``NdBSpline`` from knot vectors and a coefficient grid.
+
+    Parameters
+    ----------
+    fp : float
+        Residual sum of squares of the produced fit (kept for upstream use).
+    tck : tuple
+        Tuple ``(tx, ty, C)`` where ``tx``, ``ty`` are knot vectors and ``C``
+        is a coefficient array with shape ``(nx - kx - 1, ny - ky - 1)`` or
+        a compatible shape that can be reshaped to that.
+    degrees : tuple of int
+        Degrees ``(kx, ky)`` along x and y.
+
+    Returns
+    -------
+    NdBSpline
+        The constructed 2D spline.
+
+    Notes
+    -----
+    Only repacks the coefficient grid; ``fp`` is not used internally here.
+    """
+    nx, ny = len(tck[0]), len(tck[1])
+    kx, ky = degrees
+    c = tck[2].reshape(nx - kx - 1, ny - ky - 1)
+    return NdBSpline((tck[0], tck[1]), c, degrees)
+
+
+def _stack_augmented_fitpack(A, offs_a, D, offs_d, nc, k, p):
+    if p == -1:
+        return A, offs_a, nc
+
+    nz = k + 1
+    AA = np.zeros((nc + D.shape[0], k+2), dtype=float)
+    AA[:nc, :nz] = A[:nc, :]
+    AA[nc:, :] = D / p
+    offset = np.r_[offs_a, offs_d]
+    return AA, offset, nc
+
+def will_square_overflow(x, dtype=np.float64):
+    """
+    Check if squaring values would overflow the given dtype.
+
+    Parameters
+    ----------
+    x : array_like
+        Input values to test.
+    dtype : dtype, optional
+        Floating dtype to test against, by default ``np.float64``.
+
+    Returns
+    -------
+    ndarray of bool
+        Boolean mask where True indicates ``x**2`` would overflow ``dtype``.
+
+    Notes
+    -----
+    Compares ``abs(x)`` to ``sqrt(np.finfo(dtype).max)`` elementwise.
+    """
+    max_val = np.finfo(dtype).max
+    return abs(x) > np.sqrt(max_val)
+
+def fp_residual(Z, Zhat, W=None):
+    """
+    Compute FITPACK-style residual sum of squares ``fp``.
+
+    Parameters
+    ----------
+    Z : array_like
+        Target data (observed grid).
+    Zhat : array_like
+        Predicted/fitted grid, broadcastable to ``Z``.
+    W : array_like, optional
+        Optional weights, same shape as ``Z``; if provided, compute
+        ``sum(W * (Z - Zhat)**2)``.
+
+    Returns
+    -------
+    float
+        Residual sum of squares, or ``inf`` if non-finite or overflow detected.
+
+    Notes
+    -----
+    Guards against NaNs/Infs and overflow by early returning ``inf``.
+    """
+    R = Z - Zhat
+
+    if not np.isfinite(R).all():
+        return float("inf")
+
+    if np.isnan(R).any() or will_square_overflow(R).any():
+        return float("inf")
```

Comment:
These checks look out of place this deep into the code:
1. if `np.isfinite(R).all()` is True, then there's no need for `np.isnan(R)` just after it?
2. is there a test case where `R**2` overflows or is it theoretical? If there is, how does FITPACK handle it?